src/org/exist/storage/XQueryPool.java

```diff
@@ -146,23 +147,19 @@ private void returnObject(final Source source, final CompiledXQuery xquery) {
             if (pool.size() < maxPoolSize) {
                 source.setCacheTimestamp(System.currentTimeMillis());
 
-                final Deque<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ArrayDeque<>());
+                final Set<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ConcurrentSkipListSet<>());
 
                 if (stack.size() < maxStackSize) {
-                    // check if the query is already in pool before adding,
-                    // may happen for modules, don't add it a second time!
-                    if(!stack.contains(xquery)) {
-                        stack.push(xquery);
-                    }
+                    stack.add(xquery);
```

Comment:
> you have a race condition between checking validity and removing the query, so you may remove a query which is actually valid.

I do not see race condition here. There is only one compiled script peek up at time and `isValid ` check for it. If one is invalid then other will not pass `isValid` checks too, no?